hilti/toolchain/include/compiler/unit.h

```diff
@@ -407,13 +410,16 @@ class Unit {
                                         const hilti::rt::filesystem::path& path);
 
     context::CacheIndex _index;                     // index for the context's module cache
+    ID _unique_id;                                  // globally unique ID for this module
     hilti::rt::filesystem::path _extension;         // AST extension, which may differ from source file
     std::optional<Node> _module;                    // root node for AST (always a `Module`), if available
     std::vector<std::weak_ptr<Unit>> _dependencies; // recorded dependencies
     std::weak_ptr<Context> _context;                // global context
     std::optional<detail::cxx::Unit> _cxx_unit;     // compiled C++ code for this unit, once available
     bool _resolved = false;                         // state of resolving the AST
     bool _requires_compilation = false;             // mark explicitly as requiring compilation to C++
+
+    static std::map<ID, unsigned int> _uid_cache; // cache storing state for generating globally unique IDs
```

Comment:
Since `ID` is hashable we could use a `std::unordered_map` here. This wouldn't cost us anything, but could give minor speedups for compilations using a lot of units.